target_chains/ethereum/sdk/solidity/PythPrice.sol

```diff
@@ -0,0 +1,438 @@
+        (uint64 otherPrice, int64 otherSign) = toUnsigned(price2.price);
+
+        uint64 midprice = (basePrice * PD_SCALE) / otherPrice;
+        int32 midpriceExpo = base.expo - price2.expo + PD_EXPO;
+
+        uint64 otherConfidencePct = (price2.conf * PD_SCALE) / otherPrice;
+        uint128 conf = ((uint128(base.conf) * PD_SCALE) / otherPrice) +
+            ((uint128(otherConfidencePct) * midprice) / PD_SCALE);
+
+        if (conf >= type(uint64).max) {
+            revert ConfidenceOverflow();
+        }
+
+        return
+            PythStructs.Price({
+                price: int64(midprice) * baseSign * otherSign,
+                conf: uint64(conf),
+                expo: midpriceExpo,
+                publishTime: Math.min(price.publishTime, price2.publishTime)
+            });
+    }
+
+    /**
+     * @dev Add two prices, propagating uncertainty.
+     * @param price The first price
+     * @param price2 The second price
+     * @return The sum of the two prices
+     */
+    function add(
+        PythStructs.Price memory price,
+        PythStructs.Price memory price2
+    ) internal pure returns (PythStructs.Price memory) {
+        if (price.expo != price2.expo) {
+            revert ExponentsMustMatch();
+        }
+
+        return
+            PythStructs.Price({
+                price: price.price + price2.price,
+                conf: price.conf + price2.conf,
+                expo: price.expo,
+                publishTime: Math.min(price.publishTime, price2.publishTime)
+            });
+    }
+
+    /**
+     * @dev Multiply a price by a constant.
+     * @param price The price
+     * @param c The constant
+     * @param e The exponent of the constant
+     * @return The resulting price
+     */
+    function cmul(
+        PythStructs.Price memory price,
+        int64 c,
+        int32 e
+    ) internal pure returns (PythStructs.Price memory) {
+        return
+            mul(
+                price,
+                PythStructs.Price({
+                    price: c,
+                    conf: 0,
+                    expo: e,
+                    publishTime: price.publishTime
+                })
+            );
+    }
+
+    /**
+     * @dev Multiply two prices, propagating uncertainty.
+     * @param price The first price
+     * @param price2 The second price
+     * @return The product of the two prices
+     */
+    function mul(
+        PythStructs.Price memory price,
+        PythStructs.Price memory price2
+    ) internal pure returns (PythStructs.Price memory) {
+        PythStructs.Price memory base = normalize(price);
+        price2 = normalize(price2);
+
+        (uint64 basePrice, int64 baseSign) = toUnsigned(base.price);
+        (uint64 otherPrice, int64 otherSign) = toUnsigned(price2.price);
+
+        uint64 midprice = basePrice * otherPrice;
+        int32 midpriceExpo = base.expo + price2.expo;
+
+        uint64 conf = base.conf * otherPrice + price2.conf * basePrice;
```

Comment:
same comment here as in `div` function: some comments to explain the math would be great